### DISCUSSION TITLE
Clear and refetch?

### DISCUSSION THREAD

**Discussion by bdrazen** [TanStack/query#331] [General] [answered]
What would be the recommended way to clear + refetch a single particular query? I'm seeing `refetch()` and `queryCache.clear()`, but the former doesn't clear the cache for the query, and the latter presumably clears the cache for **all** queries. Thanks.

**Comment by bdrazen** [comment, marked as answer]
Worked it out with:
``` 
queryCache.removeQueries(myQueryKey);
myQuery.refetch();
```
Would be nice though if `useQuery()` returned a `clearCache` function or maybe `refetch()` accepted a `clearCache` option.

**Comment by tannerlinsley** [comment]
I'll definitely consider this for future versions :)

**Reply by Rinaldo** [reply]
Have you given any more thought to a `clearCache` option? It would be nice to be able to put the query in a hard loading state when calling `refetch`

**Reply by TkDodo** [reply]
you'd want `queryClient.resetQueries` for that

**Reply by Rinaldo** [reply]
That definitely works but it would be nice not to have to reach for `queryClient`

````js
const queryKey = useMemo(() => ['query', 'key', param], [param]) // queryKey needs to be memoed so it can be used as a dep of useCallback
const { data } = useQuery({
  queryKey
  /* other options */
})
const queryClient = useQueryClient()
const refetch = useCallback(() => {
  queryClient.resetQueries({ queryKey, exact: true })
}, [queryKey])
````

That's a bit of code for a referentially stable refetch function as opposed to just being able to use the one returned by `useQuery`

**Reply by bdrazen** [reply]
@Rinaldo I wrapped it in a hook (but still fancying a nice convenience function):
```js
export const useReset = (key: QueryFilters['queryKey']) => {
  const queryClient = useQueryClient();

  return useCallback(
    () => queryClient.resetQueries({ queryKey: key, exact: true }),
    [queryClient, ...key],
  );
};
```
```js
const { data } = useQuery({
  queryKey
  /* other options */
})
const reset = useReset(queryKey);
```
